### ISSUE TITLE
drcachesim: support trace files split by thread to analyze in parallel

### ISSUE THREAD

**Issue by derekbruening** [DynamoRIO/dynamorio#3230]
This is a feature proposal to support trace files stored on disk in a separate file per thread.  For very large traces this can help solve disk size limits.  It also enables parallelizing trace analysis, for tools that want to perform per-thread computations (i.e., not the cache simulator, but most of the other tools like reuse_distance (#3227) or basic_counts or opcode_mix).

One complexity is how to handle aggregation: these tools want to report whole-process results.  Rather than running a separate instance of each tool per thread, we probably want each parallel-enabled tool to split its stats by thread and support handle_memref() being called concurrently, with aggregation in print_results().

For using a split-file trace with something like the cache simulator, we could extend the trace reader to support interleaving the threads into a single timestamp-ordered stream.

**Comment by derekbruening** [comment]
Documenting design decisions:

+ We will change raw2trace to produce a separate processed file per thread in a trace/ subdir (peer of raw/).

+ We will move rawtrace's thread interleaving in timestamp order into file_reader_t.

+ We will leave support for reading legacy already-interleaved single-file traces, but stop producing them.

+ For analysis tools, parallelism will be driven by the front end, as not all tools will necessarily support it.  Parallel analysis will involve a single instance of the tool with a separate analysis thread per trace thread.  The tool must provide thread dispatch (e.g., a hashtable to split the data), thread safety, and aggregation (it is not easy to come up with a general aggregation interface that works for all tools).

+ Rawtrace can now be parallelized.  I will file a separate issue on further optimizations there: apparently the hashtable lock for the decode cache is a bottleneck.

**Comment by derekbruening** [comment]
This was finished and I think I just forgot to close it.  Any further improvements or issues are filed separately.
